File: hu.bme.mit.theta.core/src/main/java/hu/bme/mit/theta/core/type/inttype/IntExprs.java
/*
 *  Copyright 2017 Budapest University of Technology and Economics
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  
 *      http://www.apache.org/licenses/LICENSE-2.0
 *  
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.inttype;

import com.google.common.collect.ImmutableList;

import hu.bme.mit.theta.core.type.Expr;

public final class IntExprs {

	private IntExprs() {
	}

	public static IntType Int() {
		return IntType.getInstance();
	}

	public static IntLitExpr Int(final int value) {
		return new IntLitExpr(value);
	}

	public static IntToRatExpr ToRat(final Expr<IntType> op) {
		return new IntToRatExpr(op);
	}

	public static IntAddExpr Add(final Iterable<? extends Expr<IntType>> ops) {
		return new IntAddExpr(ops);
	}

	public static IntSubExpr Sub(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntSubExpr(leftOp, rightOp);
	}

	public static IntNegExpr Neg(final Expr<IntType> op) {
		return new IntNegExpr(op);
	}

	public static IntMulExpr Mul(final Iterable<? extends Expr<IntType>> ops) {
		return new IntMulExpr(ops);
	}

	public static IntDivExpr Div(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntDivExpr(leftOp, rightOp);
	}

	public static ModExpr Mod(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new ModExpr(leftOp, rightOp);
	}

	public static RemExpr Rem(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new RemExpr(leftOp, rightOp);
	}

	public static IntEqExpr Eq(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntEqExpr(leftOp, rightOp);
	}

	public static IntNeqExpr Neq(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntNeqExpr(leftOp, rightOp);
	}

	public static IntLtExpr Lt(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntLtExpr(leftOp, rightOp);
	}

	public static IntLeqExpr Leq(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntLeqExpr(leftOp, rightOp);
	}

	public static IntGtExpr Gt(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntGtExpr(leftOp, rightOp);
	}

	public static IntGeqExpr Geq(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		return new IntGeqExpr(leftOp, rightOp);
	}

	/*
	 * Convenience methods
	 */

	public static IntAddExpr Add(final Expr<IntType> op1, final Expr<IntType> op2) {
		return Add(ImmutableList.of(op1, op2));
	}

	public static IntAddExpr Add(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3) {
		return Add(ImmutableList.of(op1, op2, op3));
	}

	public static IntAddExpr Add(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3,
			final Expr<IntType> op4) {
		return Add(ImmutableList.of(op1, op2, op3, op4));
	}

	public static IntAddExpr Add(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3,
			final Expr<IntType> op4, final Expr<IntType> op5) {
		return Add(ImmutableList.of(op1, op2, op3, op4, op5));
	}

	////

	public static IntMulExpr Mul(final Expr<IntType> op1, final Expr<IntType> op2) {
		return Mul(ImmutableList.of(op1, op2));
	}

	public static IntMulExpr Mul(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3) {
		return Mul(ImmutableList.of(op1, op2, op3));
	}

	public static IntMulExpr Mul(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3,
			final Expr<IntType> op4) {
		return Mul(ImmutableList.of(op1, op2, op3, op4));
	}

	public static IntMulExpr Mul(final Expr<IntType> op1, final Expr<IntType> op2, final Expr<IntType> op3,
			final Expr<IntType> op4, final Expr<IntType> op5) {
		return Mul(ImmutableList.of(op1, op2, op3, op4, op5));
	}

}

File: subprojects/common/core/src/main/kotlin/hu/bme/mit/theta/core/type/inttype/IntExprs.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.inttype

import hu.bme.mit.theta.core.type.Expr
import java.math.BigInteger

/** Factory and utility methods for integer expressions. */
@Suppress("FunctionName")
object IntExprs {
  @JvmStatic fun Int() = IntType

  @JvmStatic fun Int(value: Int) = IntLitExpr(BigInteger.valueOf(value.toLong()))

  @JvmStatic fun Int(value: String) = IntLitExpr(BigInteger(value))

  @JvmStatic fun Int(value: BigInteger) = IntLitExpr(value)

  @JvmStatic fun ToRat(op: Expr<IntType>) = IntToRatExpr(op)

  @JvmStatic fun Add(ops: Iterable<Expr<IntType>>) = IntAddExpr.of(ops)

  @JvmStatic @SafeVarargs fun Add(vararg ops: Expr<IntType>) = IntAddExpr(ops.asList())

  @JvmStatic fun Sub(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntSubExpr(leftOp, rightOp)

  @JvmStatic fun Pos(op: Expr<IntType>) = IntPosExpr(op)

  @JvmStatic fun Neg(op: Expr<IntType>) = IntNegExpr(op)

  @JvmStatic fun Mul(ops: Iterable<Expr<IntType>>) = IntMulExpr.of(ops)

  @JvmStatic @SafeVarargs fun Mul(vararg ops: Expr<IntType>) = IntMulExpr(ops.asList())

  @JvmStatic fun Div(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntDivExpr(leftOp, rightOp)

  @JvmStatic fun Mod(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntModExpr(leftOp, rightOp)

  @JvmStatic fun Rem(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntRemExpr(leftOp, rightOp)

  @JvmStatic fun Eq(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntEqExpr(leftOp, rightOp)

  @JvmStatic fun Neq(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntNeqExpr(leftOp, rightOp)

  @JvmStatic fun Lt(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntLtExpr(leftOp, rightOp)

  @JvmStatic fun Leq(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntLeqExpr(leftOp, rightOp)

  @JvmStatic fun Gt(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntGtExpr(leftOp, rightOp)

  @JvmStatic fun Geq(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntGeqExpr(leftOp, rightOp)
}
